Connectivity car and bug fixing

shonaighdouglas opened the pull request:
Added code to the carLogic sketch to allow it to connect to WiFi and set up a web server so the app can communicate with the car. The code used can be found [here](url). The car  can now connect to WiFi and has a static IP address.

**Edit:** carLogic wifi connectivity is now based on [another example](https://github.com/DIT112-V20/kotlin-app-arduino-sketch-ci/blob/master/arduino/SmartcarWebserver/SmartcarWebserver.ino) which also uses mDNS so we should be able to access the car by `djsmartcar.local`.

List of requests it should be able to support for now:
* `djsmartcar.local/dance?id=1` - where you can change 1 to be any integer (checks needs to be implemented)
* `djsmartcar.local/random`

The pull request also contains some bug-fixes which were done in sprint 3. 

Resolves #30, #21, #25, #39

shonaighdouglas committed:
Fix dance bugs and missing setSpeed(0)

Added setSpeed(0) to Macarena method, delays to steps of Shuffle and Shake methods, reduced speed  of dance moves.

Resolves #21, #25 and #39.

shonaighdouglas committed:
Make car connect to the internet

Used tutorial found at https://randomnerdtutorials.com/esp32-web-server-arduino-ide/ to add WiFi connectivity to carLogic sketch.

Resolves #30.

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+
+// Assign output variables to GPIO pins
+const int output26 = 26;
+const int output27 = 27;
```
Line 47-53 can be removed because it's related to the LEDs from the tutorial.

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+  pinMode(output27, OUTPUT);
+  // Set outputs to LOW
+  digitalWrite(output26, LOW);
+  digitalWrite(output27, LOW);
```
Line 60-65 can also be removed because it is related to the tutorial example.

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+            } else if (header.indexOf("GET /27/off") >= 0) {
+              Serial.println("GPIO 27 off");
+              output27State = "off";
+              digitalWrite(output27, LOW);
```
Lines: 114, 115, 118, 119, 122, 123, 126, 127
^ those lines have to be removed since we removed the variables earlier (the stuff related to the LEDs from the tutorial). 💡

shonaighdouglas replied:
I have removed the suggested lines and left the code as follows, using part of your suggestion from the next code correction. I don't know what command would be a replacement for "GET /26/off", however, so I hesitate to put anything else in there for the moment:

`            // turns the GPIOs on and off
            if (header.indexOf("GET /dance?id=1") >= 0) {
              Serial.println("Start Spinning");
            } else if (header.indexOf("GET /26/off") >= 0) {
              Serial.println("Stop Spinning");
            } else if (header.indexOf("GET /27/on") >= 0) {
              Serial.println("GPIO 27 on");
            } else if (header.indexOf("GET /27/off") >= 0) {
              Serial.println("GPIO 27 off");
            }`

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+            client.println();
+            
+            // turns the GPIOs on and off
+            if (header.indexOf("GET /26/on") >= 0) {
```
I think this (the block of if-statements) is the part that we have to change to fit the requests that will come from the app!

So I think that instead of `header.indexOf("GET /26/on")` we should be able to do e.g. `header.indexOf("GET /dance?id=1")` and then turn on one of the dance moves inside of that if-statement.

Let me know if you have any questions about this part and I can give some examples or we can code it next time we work together! 😄

shonaighdouglas replied:
I think a code-along here may be safest, and I may actually learn something from it too!

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+            client.println("<body><h1>ESP32 Web Server</h1>");
+            
+            // Display current state, and ON/OFF buttons for GPIO 26  
+            client.println("<p>GPIO 26 - State " + output26State + "</p>");
```
This line will have to be modified/removed since we're removing that variable.

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+            // Display current state, and ON/OFF buttons for GPIO 26  
+            client.println("<p>GPIO 26 - State " + output26State + "</p>");
+            // If the output26State is off, it displays the ON button       
+            if (output26State=="off") {
```
Not sure how to best do this but that variable is also removed so it won't compile, maybe remove the whole if-else part (line 146-151)?

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+            } 
+               
+            // Display current state, and ON/OFF buttons for GPIO 27  
+            client.println("<p>GPIO 27 - State " + output27State + "</p>");
```
Same as before:
Variable should be removed so this line won't work. I'd remove it because we won't need the car to give us an HTML site. 😄

JenniNord commented on src/carLogic/carLogic.ino:
```diff
@@ -23,14 +30,160 @@ DirectionalOdometer rightOdometer (smartcarlib::pins::v2::rightOdometerPins, [](
+            // Display current state, and ON/OFF buttons for GPIO 27  
+            client.println("<p>GPIO 27 - State " + output27State + "</p>");
+            // If the output27State is off, it displays the ON button       
+            if (output27State=="off") {
```
And again, removed that variable so line 155-160 can be removed or changed so the sketch compiles!

JenniNord reviewed (dismissed):
Let me know if anything was unclear, you disagree with something or have a better idea! 😄

shonaighdouglas committed:
Make suggested code edits

Removed the sections of code suggested by Jenni in the pull request review, all pertaining to the functions in the tutorial geared towards making LEDs light up. These are not required for our purposes.

Resolves #30.

JenniNord committed:
Change webserver implementation

Changed from using just WebServer.h to this example: https://github.com/DIT112-V20/kotlin-app-arduino-sketch-ci/blob/master/arduino/SmartcarWebserver/SmartcarWebserver.ino
Added in some logic for handling requests such as /random and /dance?id=1. Also changed some comments and moved the random dance move logic into a separate method for now as we're thinking of randomizing dances in the app logic and not on the car itself. As it is now, the random dance will loop 5 times which means that 5 moves will be performed by the car.

The mDNS name of the car is set to be djsmartcar and should be accessible through djsmartcar.local.

Resolves #30

shonaighdouglas commented:
Car now connects to WiFi and will perform a selected dance move or do random dances when sent one of the noted requests.

FayonaC reviewed (approved):
Updates look great!
